drivers/platform/mbed/mbed_spi.cpp

```diff
@@ -0,0 +1,280 @@
+
+	/* Allocate memory for mbed_spi_desc structure */
+	mbed_spi_desc = (struct mbed_spi_desc *)calloc(1, sizeof(*mbed_spi_desc));
+	if (!mbed_spi_desc)
+		goto err_mbed_spi_desc;
+
+	/* Configure and instantiate SPI peripheral */
+	if (((struct mbed_spi_init_param *)param->extra)->use_sw_csb) {
+		/* CSB pin toggled explicitly (s/w controlled) during SPI transaction */
+		spi = new SPI(
+			(PinName)(((struct mbed_spi_init_param *)param->extra)->spi_mosi_pin),
+			(PinName)(((struct mbed_spi_init_param *)param->extra)->spi_miso_pin),
+			(PinName)(((struct mbed_spi_init_param *)param->extra)->spi_clk_pin));
+
+		if (spi) {
+			/* Configure and instantiate chip select pin */
+			csb = new DigitalOut((PinName)(spi_desc->chip_select));
+			if (csb) {
+				mbed_spi_desc->use_sw_csb = true;
+				mbed_spi_desc->csb_gpio = (DigitalOut *)csb;
+				csb->write(GPIO_HIGH);
+			} else {
+				goto err_csb;
+			}
+		}
+	} else {
+		/* CSB pin toggled implicitly (through HAL layer) during SPI transaction */
+		spi = new SPI(
+			(PinName)(((struct mbed_spi_init_param *)param->extra)->spi_mosi_pin),
+			(PinName)(((struct mbed_spi_init_param *)param->extra)->spi_miso_pin),
+			(PinName)(((struct mbed_spi_init_param *)param->extra)->spi_clk_pin),
+			(PinName)(param->chip_select),
+			use_gpio_ssel);
+
+		mbed_spi_desc->use_sw_csb = false;
+		mbed_spi_desc->csb_gpio = NULL;
+	}
+
+	if (!spi) {
+		goto err_spi;
+	}
+
+	mbed_spi_desc->spi_port = (SPI *)spi;
+
+	spi_desc->extra = (struct mbed_spi_desc *)mbed_spi_desc;
+	*desc = spi_desc;
+
+	/**
+		NOTE: Actual frequency of SPI clk will be somewhat device
+		dependent, relating to clock-settings, prescalars etc. If absolute
+		SPI frequency is required, consult your device documentation.
+		**/
+	spi->frequency(param->max_speed_hz);
+	spi->format(SPI_8_BIT_FRAME, param->mode);   // data write/read format
+	spi->set_default_write_value(0x00);          // code to write when reading back
+
+	return SUCCESS;
+
+err_spi:
+	if (((struct mbed_spi_init_param *)param->extra)->use_sw_csb) {
+		free(csb);
+	}
+err_csb:
+	free(spi);
+	free(mbed_spi_desc);
+err_mbed_spi_desc:
+	free(spi_desc);
+
+	return -ENOMEM;
+}
+
+/**
+ * @brief Free the resources allocated by spi_init().
+ * @param desc[in, out] - The SPI descriptor.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_spi_remove(struct spi_desc *desc)
+{
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	if (((struct mbed_spi_desc *)desc->extra)->use_sw_csb) {
+		/* Free the CSB gpio object */
+		if ((DigitalOut *)(((struct mbed_spi_desc *)(desc->extra))->csb_gpio)) {
+			delete((DigitalOut *)(((struct mbed_spi_desc *)(desc->extra))->csb_gpio));
+		}
+	}
+
+	/* Free the SPI port object */
+	if ((SPI *)(((struct mbed_spi_desc *)(desc->extra))->spi_port))
+		delete((SPI *)(((struct mbed_spi_desc *)(desc->extra))->spi_port));
+
+	/* Free the SPI extra descriptor object */
+	if ((struct mbed_spi_desc *)(desc->extra))
```

Comment:
Since the driver doesn't function without the `desc->extra` parameter allocated, don't include this check. This can also be the reason for the memleak error.

Reply:
Make sense. Removed this check.